src/boot/efifirmware.c

```diff
@@ -0,0 +1,60 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+#include <endian.h>
+
+static bool efifw_validate_header(const void *blob, size_t uki_efifw_len,
+                                  uint32_t *fwid_off, uint32_t *payload_off) {
+        if ((uintptr_t) blob % alignof(EfiFwHeader) != 0)
+                return false;
+        const EfiFwHeader *fw_header = ASSERT_PTR(blob);
+
+        if (le32toh(fw_header->base.magic) != FWHEADERMAGIC)
+                return false;
+
+        size_t payload_len = le32toh(fw_header->base.payload_len);
+        size_t header_len = le32toh(fw_header->base.header_len);
+        size_t fwid_len = le32toh(fw_header->base.fwid_len);
```

Comment:
you are reading fields here that might not actually be in memory, because uki_efifw_len is smaller than the structure.